docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+   sandbox functionality (webasm, CHERI), otherwise fail and close the sandbox.
+3. When Verona code makes calls to the foreign library, the compiler-generated
+   serialisation code will transform the arguments and return value into a
+   memory buffer and call the dispatcher with the appropriate index and buffer.
+4. The dispatcher will call a templated wrapper that unpacks the arguments and
+   calls the actual function.
+5. The return value, if any, is desserialised and bound to the Verona variable
+   or used as an argument in a Verona call.
+
+## C/C++ interoperability
+
+For both C and C++, we use `clang` to parse the header file to know what types
+are declared, functions are exported, etc. This allows us to know which of the
+Verona code lines calling foreign code are correct, and generate the correct
+serialisation routines and, for C++, instantiating the correct template
+implementations based on the types used in Verona.
+
+The C++ interoperability layer parses the file and generates the foreign AST.
+We then use the Verona code function names and types to create
+a function declaration (context, name, argument types) and, if valid, we add
+that function to the dispatcher, giving it its own (auto-incremented) index.
+
+Additional code needs to be generated to serialise the Verona arguments and
+return values into a buffer, so that the dispatcher can be generic, only taking
+an index and a memory buffer.
+
+For C++, further code will be created (via AST construction and LLVM code
+generation) if the types involved are parametrised. All template arguments are
+taken into consideration and a specialisation is defined and generated in a
+wrapper layer.
+
+## Child process execution environment
+
+Our initial implementation is to separate the sandbox execution by creating a
+child process. This is not strictly speaking a sandbox, but gives us an easy
+way to test the rest of the framework without much additional work. Our aim is
+to use WebAssembly as our default sandbox environment, but that will take time
+and will be done later.
+
+The sandbox creation forks the process with a special sandbox library that will
+forward all calls to the parent process, which will control where the memory
+actually lives (see below). A local heap is created per sandbox for the child
+to use in its allocations.
+
+Calls will be forwarded across the parent/child barrier via the dispatcher, by
+means of a function pointer table, where the position is the index in the code
+generation stage above, and the buffer is the contents of the arguments and a
+place for the return value.
+
+Argument decomposition will be done by the sandbox code that will call the actual
+function with the actual arguments and, if there is a return value, set it on
+the right place of the buffer and return.
+
+### Memory Management
+
+To avoid the sandbox from requesting memory in the wrong places (accidentally
```

Comment:
This is... not a very close approximation of what's going on.

For integer-pointer environments (processes, SFI) we carve out a chunk of the virtual address space for the shared heap.  We use some explicit calls to transfer ownership of large regions within this (superslabs) between the Verona code and the sandbox and then both sides have their own snmalloc `Alloc` instances that are able to allocate memory.  We don't need an RPC per malloc, that would be prohibitively expensive.  We started by making malloc across security contexts cheap and then built a lot of the other mechanisms on top of this.